### PR TITLE
Upvote / Downvote questions

### PR DESCRIPTION
**Pull request by filip-gregurevic**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Currently students have no possibility to show that they are having the same question or to rate questions in general.

### Description

Students can upvote / downvote questions to express (dis-)agreement.
There are no negative votes. If the vote is at 0 a student cannot downvote the question.
Questions are ordered by their votes.

### Steps for Testing

1. Log in to Artemis
(1a. Enable questions and answers for your test course if not already enabled) 
2. Navigate to an exercise or lecture
3. Create multiple questions
4. Upvote / Downvote questions in any order
5. Check if sorting and representation works correctly

### Screenshots

default:
Bildschirmfoto 2020-05-25 um 18 21 48

upvoted:
Bildschirmfoto 2020-05-18 um 18 37 31
downvoted:
Bildschirmfoto 2020-05-25 um 18 23 31

### CODE REVIEW

**Comment by filip-gregurevic** [comment]
> 1. Are we sure that we want to have downvotes? This might make students hesitate to ask any questions since they could be afraid of being downvoted heavily. I get the idea of using downvoting as a mechanism to hide inappropriate questions but then this would need to be solved differently than it is right now (and in that case deleting the question all together might even be the preferred cause of action). An alternative solution would be to allow downvoting but to not expose any score of less than zero, meaning zero points is the floor that will be visibily shown (and returned from the server). That way downvoting still has a purpose but no one is publicly humiliated by having his or her name next to 100 downvotes.

Yes, I thought about this point a lot, too. I think deleting that question automatically is a bit too harsh and should only be done manually by a tutor or instructor. 
I like the idea of capping the votes at zero to prevent humiliation. @krusche What do you think about this idea?

> 2. From a technical perspective, the voting is very easy to manipulate at the moment since it is not enforced that one user can only cast a single vote per question. It would be pretty easy to inspect the HTTP request that happens on upvoting and then making the same request again this time with a value of 10,000 to give that many upvotes or downvotes to a question. I can definitely see this happening when a whole bunch of CS students discover this new feature and some of them wonder if they can tinker with it.

I discussed this point with @krusche and we decided to start with this implementation since we do not assume any malicious intent by the students. (They could also upvote / downvote a question unlimited times if they clear their browser cache). But if this becomes a problem, I have the very secure (but also more complicated and performance affecting) implementation planned.

**Comment by madwau** [comment]
> Yes, I thought about this point a lot, too. I think deleting that question automatically is a bit too harsh and should only be done manually by a tutor or instructor.
> I like the idea of capping the votes at zero to prevent humiliation. @krusche What do you think about this idea?

I was not suggesting any sort of automatic deletion, that would indeed be pretty drastic.

> I discussed this point with @krusche and we decided to start with this implementation since we do not assume any malicious intent by the students. (They could also upvote / downvote a question unlimited times if they clear their browser cache).

I imagined that this was probably a conscious decision to keep the implementation effort reasonably low for a first version of the feature. Maybe we could at least add a boundary check to the server endpoint so that only deltas of ±2 are allowed? That way it's at least a little harder to completely screw with the votes for a question. Of course students could still send hundreds of requests to achieve the same effect but I think that sets a higher bar since that would basically be a spamming attack which is more severe than taking advantage of a loophole.

**Comment by filip-gregurevic** [comment]
@madwau I now added the check at the endpoint and the displayed votes are also now capped at zero. They can't be downvoted if at zero, only a downvote can be _reset_ if negative or at zero.
Thanks again for the suggestions! 😄

**Comment by madwau** [comment]
> @madwau I now added the check at the endpoint and the displayed votes are also now capped at zero. They can't be downvoted if at zero, only a downvote can be _reset_ if negative or at zero.
> Thanks again for the suggestions!

Thanks for the adaptations! For the zero flooring, I was thinking more of a server-side solution so that negative scores are not exposed to users at all.

File: src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java [hunk 1]
```diff
@@ -214,4 +214,41 @@ public void deleteStudentQuestions_asStudent() throws Exception {
...
+    public void editStudentQuestionVotes_asStudent() throws Exception {
+        List<StudentQuestion> questions = database.createCourseWithExerciseAndStudentQuestions();
+        StudentQuestion studentQuestion = questions.get(0);
+
+        studentQuestion.setVotes(42);
+
+        StudentQuestion updatedStudentQuestion = request.putWithResponseBody("/api/student-questions/" + studentQuestion.getId() + "/votes", 0, StudentQuestion.class,
+                HttpStatus.OK);
+        assertThat(updatedStudentQuestion.getVotes().equals(42));
+    }
```
Comment (sleiss) [inline src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java:253]: I believe some test cases where the value actually gets changed would be quite useful.

Comment (filip-gregurevic) [reply src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java:247]: These were just placeholders, now there are tests that actually test something with https://github.com/ls1intum/Artemis/pull/1478/commits/45b799f736ae7a4a3ae31813144fb489be783d9b

**Comment by filip-gregurevic** [comment]
> The functionality works as intended, the only thing that I'm not really sure about is if this is intended: There seems to be a different behaviour if you remove an upvote or downvote, this was a bit confusing for me, especially as the vote count is limited as 0.

That behaviour is only around the 0 value, which is intended since the votes are capped at 0

File: src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java (hunk 1 above)
Comment (sleiss) [reply src/test/java/de/tum/in/www1/artemis/StudentQuestionIntegrationTest.java:247]: Yeah, thats how I like them test cases :D

**Comment by filip-gregurevic** [comment]
> Hey Filip, I am not sure but I think I found a bugs🐇
> When I up vote from 0 it turns to 2 directly and skips 1 as shown in the screen cast
@HanyaElhashemy thank you for testing 😃 
In the screen cast it seems to be intended behaviour. Since you take back your downvote and upvote it simultaneously the total should increase by 2

**Pull request merged**